### ISSUE TITLE
Feature Request: AppService Environments (ASE) Support

### ISSUE THREAD

**Issue by mjyeaney** [terraform-providers/terraform-provider-azurerm#438]
Not an issue; rather, looking for support of ASE deployments. Any pointers / timelines on status would be awesome.

**Comment by tombuildsstuff** [comment]
hey @mjyeaney 

Thanks for opening this issue - apologies for the delayed response here!

I've been taking a look into how we'd support App Service Environments this afternoon and have opened #869 which is a WIP PR adding support for App Service Environments into Terraform. In order to validate the Schema is correct (and not missing any commonly used fields), would it be possible to find out which features of App Service Environments you're using?

Thanks!

**Comment by mjyeaney** [comment]
No problem - the basic customer scenarios for deploying ASE's are essentially the following:

* One or more back-end worker pools
* Support both [internal](https://docs.microsoft.com/en-us/azure/app-service/environment/create-ilb-ase) or [external](https://docs.microsoft.com/en-us/azure/app-service/environment/create-external-ase) deployment modes.
* Specification of [SSL certificates](https://docs.microsoft.com/en-us/azure/app-service/environment/create-from-template#upload-and-configure-the-default-ssl-certificate)

Definitely appreciate the effort here - let me know if you need additional clarification.

Thanks!

**Comment by achandmsft** [comment]
@tombuildsstuff does this still have an upstream dependency?

**Comment by pietro** [comment]
@achandmsft #917 is IMHO an upstream dependency.

**Comment by rifaterdemsahin** [comment]
what is the state on this one ?

**Comment by tombuildsstuff** [comment]
Unfortunately the App Service Environment API is still broken - as such this is on hold until that's fixed (which is why 
https://github.com/terraform-providers/terraform-provider-azurerm/pull/869 is currently closed).

**Comment by pietro** [comment]
@achandmsft I filled Azure/azure-rest-api-specs#3648 and created Azure/azure-rest-api-specs#3774 so hopefully this can get fixed at the Azure API swagger and then pushed to the go SDK which help with the progress on this issue.

**Comment by tombuildsstuff** [comment]
@pietro thanks, unfortunately this blocking issue at this time is where App Service Environment doesn't finish provisioning (we stopped checking after 30 hours) for both an ASEv1 and an ASEv2 - which we couldn't find a way around - there's a thread tracking that (but I don't recall if it's on the Rest API Specs repo or through an internal channel); let me push for an update :)

**Comment by MarcoMol** [comment]
Hi @tombuildsstuff , did you get an update on that push?
I'm quite curious on the status as our company has put ASEv2 on the roadmap as a possibility for landing SaaS applications. :)

**Comment by mrwalters1988** [comment]
Hey all, Also hoping for an update to this. We're trying to transition over to Terraform and this is a blocker currently.

**Comment by Midacts** [comment]
We run one ASE in production. I thought this as well when I started. I just
keep an eye on it and know that right now we won't configure the ASE via
terraform. No biggie.

**Comment by pricoppd** [comment]
Any updates on this?

**Comment by liamfoneill** [comment]
I would use ARM templates for now directly as even executing the ARM template with Terraform runs afoul of the 1 hour timeout bug.


-------- Original message --------

**Comment by pricoppd** [comment]
I use ARM templates and Terraform to deploy ASE, but if you a have a  pipeline that starts with an ASE deployment the rest of the steps will fail (deploying on top - app service plans, webapps, function apps etc) because of the timeout.
i can add a step to wait ~40 minutes to make sure the ASE is deployed but i don't really like this...

**Comment by geglne** [comment]
Any updates on this? I have deployed 6-10 of these over the last 3-4 months. As we all know, this takes a long time and therefore I can't use Terraform to deploy these :(.

I have also provided feedback to MSFT.

**Comment by KevinDMack** [comment]
ARM is a path forward but am working with several groups on this that would prefer it be included within Terraform proper.  The problem with the ARM Templates is that the graph is not incorporated into the deployment which complicates the end deployment.

**Comment by logachev** [comment]
@tombuildsstuff I saw your issues filed on the github for ASE provisioning.

I was trying to run your PR to investigate it a little bit more. I found multiple problems with existing REST APIs and I filed bugs for the ASE team using internal feedback channel, also I sent issue details to the PM, responsible for this service.

### Create issue
https://github.com/Azure/azure-rest-api-specs/issues/4915 -- this issue is not a provisioning issue, it is a REST API spec problem. The issue is that spec lists error codes as an expected responses without error details, so go SDK gets '400 Bad Request', but keeps retrying.

See https://github.com/logachev/azure-rest-api-specs/pull/1 for the proposed fix I shared with ASE team.

When this issue is resolved, ASE provisioning via the Terraform works (with a few minor changes in the old PR, I can share it later). (Tested with custom generated golang SDK)

### Delete issue
Delete API has 2 issues:
  - Delete request operation status URI lives under ASE resource and it leads to the racing between SDK gets 200 or 404 (because ASE resource is removed). If SDK gets 200 everything is fine, if it gets 404 it has a long running retry logic that fails after some time.
  - Delete operation returns success before all resources are actually removed. They're no longer available or visible via the portal, but you can't remove VNET because subnet is still in use.

I will post updates when I have them, but it is outside of my control :(

**Comment by ctooley21** [comment]
any update?

**Comment by logachev** [comment]
As far as I know, Product Group was targeting July to release fixes.

**Comment by nexxai** [comment]
@logachev Just wondering if the fixes are still on track for release this month?

**Comment by logachev** [comment]
@nexxai sorry for the late response. I sent an email to the PG to ask for an ETA.

Around 10 days ago I was asked to help PG to verify if their fixes unblock Terraform ASE provisioning. I will try to find some time next week to update @tombuildsstuff old ASE branch with latest SDKs and check if problem goes away.

**Comment by ctooley21** [comment]
any update?

**Comment by logachev** [comment]
@ctooley21 sorry, I didn't get any updates yet :(

**Comment by egorchabala** [comment]
So what's the current status? Still nothing from PG?

**Comment by logachev** [comment]
@tombuildsstuff Azure GO sdk released `v34.2.0` which includes latest ASE SDK. Do you have an ETA when the plugin can be updated to this version?

**Comment by katbyte** [comment]
@logachev, we are now on v35 of the sdk in master 🙂

**Comment by logachev** [comment]
Finally have some good news.
I have a rough implementation working for ASE v2 (not sure what is actually the status of ASE v1 and if it is needed at all).
I didn't test a lot of edge cases, but some defaults are working. 

If you can share sample configurations you're interested in I can give it a try.
It takes forever to Create, so will try to send a PR in a few days.

**Comment by schlbra** [comment]
Any updates on the feature? This would be extremely helpful to our organization.

**Comment by collinstevens** [comment]
@logachev here are some basic sample configuration templates we use for our ARM templates right now, https://github.com/Azure/azure-quickstart-templates/tree/master/201-web-app-asev2-ilb-create/ and https://github.com/Azure/azure-quickstart-templates/tree/master/201-web-app-asev2-create/. I would much rather be able to manage these through Terraform.
